python/lsst/daf/butler/tests/_examplePythonTypes.py

```diff
@@ -25,8 +25,113 @@
+    `MetricsExample` objects.
+
+    This method allows `MetricsExample` to be used with test repositories
+    in any package without needing to provide a custom configuration there.
+
+    Parameters
+    ----------
+    butler : `lsst.daf.butler.Butler`
+        The repository that needs to support `MetricsExample`.
+
+    Notes
+    -----
+    This method enables the following storage classes:
+
+    ``StructuredData``
+        A `MetricsExample` whose ``summary``, ``output``, and ``data`` members
+        can be retrieved as dataset components.
+    ``StructuredDataNoComponents``
+        A monolithic write of a `MetricsExample`.
+    """
+    yamlDict = _addFullStorageClass(
+        butler,
+        "StructuredDataDictYaml",
+        "lsst.daf.butler.formatters.yamlFormatter.YamlFormatter",
+        pytype=dict,
+    )
+
+    yamlList = _addFullStorageClass(
+        butler,
+        "StructuredDataListYaml",
+        "lsst.daf.butler.formatters.yamlFormatter.YamlFormatter",
+        pytype=list,
+        parameters={"slice"},
+        assembler="lsst.daf.butler.tests.ListAssembler"
+    )
+
+    _addFullStorageClass(
+        butler,
+        "StructuredDataNoComponents",
+        "lsst.daf.butler.formatters.pickleFormatter.PickleFormatter",
+        pytype=MetricsExample,
+        parameters={"slice"},
+        assembler="lsst.daf.butler.tests.MetricsAssembler"
+    )
+
+    _addFullStorageClass(
+        butler,
+        "StructuredData",
+        "lsst.daf.butler.formatters.yamlFormatter.YamlFormatter",
+        pytype=MetricsExample,
+        components={"summary": yamlDict,
+                    "output": yamlDict,
+                    "data": yamlList,
+                    },
+        assembler="lsst.daf.butler.tests.MetricsAssembler"
+    )
+
+
+def _addFullStorageClass(butler, name, formatter, *args, **kwargs):
+    """Create a storage class-formatter pair in a repository if it does not
+    already exist.
+
+    Parameters
+    ----------
+    butler : `lsst.daf.butler.Butler`
+        The repository that needs to contain the class.
+    name : `str`
+        The name to use for the class.
+    formatter : `str`
+        The formatter to use with the storage class. Ignored if ``butler``
+        does not use formatters.
+    *args
+    **kwargs
+        Arguments, other than ``name``, to the `~lsst.daf.butler.StorageClass`
+        constructor.
+
+    Returns
+    -------
+    The newly created storage class, or the class of the same name previously
+    found in the repository.
+    """
+    storageRegistry = butler.datastore.storageClassFactory
+
+    storage = StorageClass(name, *args, **kwargs)
+    try:
+        storageRegistry.registerStorageClass(storage)
+    except ValueError:
+        storage = storageRegistry.getStorageClass(name)
+
+    try:
+        # if lsst.daf.butler.datastores.fileLikeDatastore.FileLikeDatastore
+        formatterRegistry = butler.datastore.formatterFactory
```

Comment:
I know it's unpythonic, but might it be worth adding an abstract method to `Datastore` that does this (and raises, of course, if there is no formatter at all)?

Reply:
Add this to `datastore.py`:
```python
@property
def datastores(self):
    return [self]
```
?

Reply:
I'm trying to write a regression test for this problem. Is there an easy way to get a `ChainedDatastore` repository that doesn't support `MetricsExample` (so that I can see if `registerMetricsExample` registers the formatter correctly)?

Reply:
I think the easiest is to copy the chained datastore yaml test config and stop it loading the metrics formatters. Then use that.

Reply:
I ended up going with the original double-`try` approach. Adding a property to `Datastore` broke `ABCDataStore`.